File: artifacts/api-server/src/routes/textbooks.ts
import { Router, type IRouter } from "express";
import { eq, count, asc } from "drizzle-orm";
import { db, textbooksTable, unitsTable, wordsTable } from "@workspace/db";
import {
  CreateTextbookBody,
  UpdateTextbookParams,
  UpdateTextbookBody,
  DeleteTextbookParams,
  GetTextbookParams,
  ListTextbooksResponse,
  GetTextbookResponse,
  UpdateTextbookResponse,
  GetStatsSummaryResponse,
} from "@workspace/api-zod";
import { sd } from "../lib/serialize";
const router: IRouter = Router();

router.get("/textbooks", async (_req, res): Promise<void> => {
  const textbooks = await db.select().from(textbooksTable).orderBy(asc(textbooksTable.orderIndex), asc(textbooksTable.createdAt));
  res.json(ListTextbooksResponse.parse(textbooks.map(sd)));
});

router.post("/textbooks", async (req, res): Promise<void> => {
  const parsed = CreateTextbookBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }
  const existing = await db.select().from(textbooksTable).orderBy(asc(textbooksTable.orderIndex), asc(textbooksTable.createdAt));
  const maxIndex = existing.length > 0 ? Math.max(...existing.map((t) => t.orderIndex)) : -1;
  const [textbook] = await db.insert(textbooksTable).values({ ...parsed.data, orderIndex: maxIndex + 1 }).returning();
  res.status(201).json(GetTextbookResponse.parse(sd(textbook!)));
});

router.post("/textbooks/reorder", async (req, res): Promise<void> => {
  const { items } = req.body as { items?: { id: number; orderIndex: number }[] };
  if (!Array.isArray(items)) { res.status(400).json({ error: "items required" }); return; }
  for (const item of items) {
    await db.update(textbooksTable).set({ orderIndex: item.orderIndex }).where(eq(textbooksTable.id, item.id));
  }
  res.json({ ok: true });
});

router.get("/textbooks/:id", async (req, res): Promise<void> => {
  const params = GetTextbookParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }
  const [textbook] = await db.select().from(textbooksTable).where(eq(textbooksTable.id, params.data.id));
  if (!textbook) {
    res.status(404).json({ error: "Textbook not found" });
    return;
  }
  res.json(GetTextbookResponse.parse(sd(textbook)));
});

router.patch("/textbooks/:id", async (req, res): Promise<void> => {
  const params = UpdateTextbookParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }
  const parsed = UpdateTextbookBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }
  const [textbook] = await db
    .update(textbooksTable)
    .set({ ...parsed.data, updatedAt: new Date() })
    .where(eq(textbooksTable.id, params.data.id))
    .returning();
  if (!textbook) {
    res.status(404).json({ error: "Textbook not found" });
    return;
  }
  res.json(UpdateTextbookResponse.parse(sd(textbook)));
});

router.delete("/textbooks/:id", async (req, res): Promise<void> => {
  const params = DeleteTextbookParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }
  const [textbook] = await db.delete(textbooksTable).where(eq(textbooksTable.id, params.data.id)).returning();
  if (!textbook) {
    res.status(404).json({ error: "Textbook not found" });
    return;
  }
  res.sendStatus(204);
});

router.get("/stats/summary", async (_req, res): Promise<void> => {
  const [textbooksCount] = await db.select({ count: count() }).from(textbooksTable);
  const [unitsCount] = await db.select({ count: count() }).from(unitsTable);
  const [wordsCount] = await db.select({ count: count() }).from(wordsTable);
  res.json(
    GetStatsSummaryResponse.parse({
      totalTextbooks: Number(textbooksCount?.count ?? 0),
      totalUnits: Number(unitsCount?.count ?? 0),
      totalWords: Number(wordsCount?.count ?? 0),
    }),
  );
});

export default router;

File: python-word-quiz/api/routes/textbooks.py
from flask import Blueprint, jsonify, request
from db import get_db

bp = Blueprint('textbooks', __name__)

@bp.route('/api/textbooks', methods=['GET'])
def get_textbooks():
    db = get_db()
    textbooks = db.execute('''
    SELECT * FROM textbooks ORDER BY order_index ASC, created_at ASC
    ''').fetchall()
    
    result = []
    for textbook in textbooks:
        result.append({
            'id': textbook['id'],
            'name': textbook['name'],
            'description': textbook['description'],
            'orderIndex': textbook['order_index'],
            'createdAt': textbook['created_at'],
            'updatedAt': textbook['updated_at']
        })
    
    return jsonify(result)

@bp.route('/api/textbooks', methods=['POST'])
def create_textbook():
    data = request.get_json()
    if not data or 'name' not in data:
        return jsonify({'error': 'Name is required'}), 400
    
    db = get_db()
    
    # 获取当前最大的order_index
    max_index = db.execute('''
    SELECT COALESCE(MAX(order_index), -1) as max_index FROM textbooks
    ''').fetchone()['max_index']
    
    try:
        db.execute('''
        INSERT INTO textbooks (name, description, order_index)
        VALUES (?, ?, ?)
        ''', (data['name'], data.get('description'), max_index + 1))
        db.commit()
        
        # 获取新创建的教材
        new_textbook = db.execute('''
        SELECT * FROM textbooks WHERE id = last_insert_rowid()
        ''').fetchone()
        
        return jsonify({
            'id': new_textbook['id'],
            'name': new_textbook['name'],
            'description': new_textbook['description'],
            'orderIndex': new_textbook['order_index'],
            'createdAt': new_textbook['created_at'],
            'updatedAt': new_textbook['updated_at']
        }), 201
    except Exception as e:
        db.rollback()
        return jsonify({'error': str(e)}), 500

@bp.route('/api/textbooks/reorder', methods=['POST'])
def reorder_textbooks():
    data = request.get_json()
    if not data or 'items' not in data:
        return jsonify({'error': 'Items are required'}), 400
    
    db = get_db()
    try:
        for item in data['items']:
            db.execute('''
            UPDATE textbooks SET order_index = ? WHERE id = ?
            ''', (item['orderIndex'], item['id']))
        db.commit()
        return jsonify({'ok': True})
    except Exception as e:
        db.rollback()
        return jsonify({'error': str(e)}), 500

@bp.route('/api/textbooks/<int:id>', methods=['GET'])
def get_textbook(id):
    db = get_db()
    textbook = db.execute('''
    SELECT * FROM textbooks WHERE id = ?
    ''', (id,)).fetchone()
    
    if not textbook:
        return jsonify({'error': 'Textbook not found'}), 404
    
    return jsonify({
        'id': textbook['id'],
        'name': textbook['name'],
        'description': textbook['description'],
        'orderIndex': textbook['order_index'],
        'createdAt': textbook['created_at'],
        'updatedAt': textbook['updated_at']
    })

@bp.route('/api/textbooks/<int:id>', methods=['PATCH'])
def update_textbook(id):
    data = request.get_json()
    if not data:
        return jsonify({'error': 'Data is required'}), 400
    
    db = get_db()
    textbook = db.execute('''
    SELECT * FROM textbooks WHERE id = ?
    ''', (id,)).fetchone()
    
    if not textbook:
        return jsonify({'error': 'Textbook not found'}), 404
    
    try:
        update_data = []
        update_fields = []
        
        if 'name' in data:
            update_fields.append('name = ?')
            update_data.append(data['name'])
        if 'description' in data:
            update_fields.append('description = ?')
            update_data.append(data['description'])
        
        update_data.append(id)
        
        db.execute(f'''
        UPDATE textbooks SET {', '.join(update_fields)}, updated_at = CURRENT_TIMESTAMP
        WHERE id = ?
        ''', update_data)
        db.commit()
        
        updated_textbook = db.execute('''
        SELECT * FROM textbooks WHERE id = ?
        ''', (id,)).fetchone()
        
        return jsonify({
            'id': updated_textbook['id'],
            'name': updated_textbook['name'],
            'description': updated_textbook['description'],
            'orderIndex': updated_textbook['order_index'],
            'createdAt': updated_textbook['created_at'],
            'updatedAt': updated_textbook['updated_at']
        })
    except Exception as e:
        db.rollback()
        return jsonify({'error': str(e)}), 500

@bp.route('/api/textbooks/<int:id>', methods=['DELETE'])
def delete_textbook(id):
    db = get_db()
    textbook = db.execute('''
    SELECT * FROM textbooks WHERE id = ?
    ''', (id,)).fetchone()
    
    if not textbook:
        return jsonify({'error': 'Textbook not found'}), 404
    
    try:
        db.execute('''
        DELETE FROM textbooks WHERE id = ?
        ''', (id,))
        db.commit()
        return '', 204
    except Exception as e:
        db.rollback()
        return jsonify({'error': str(e)}), 500

@bp.route('/api/stats/summary', methods=['GET'])
def get_stats_summary():
    db = get_db()
    
    textbooks_count = db.execute('''
    SELECT COUNT(*) as count FROM textbooks
    ''').fetchone()['count']
    
    units_count = db.execute('''
    SELECT COUNT(*) as count FROM units
    ''').fetchone()['count']
    
    words_count = db.execute('''
    SELECT COUNT(*) as count FROM words
    ''').fetchone()['count']
    
    return jsonify({
        'totalTextbooks': textbooks_count,
        'totalUnits': units_count,
        'totalWords': words_count
    })